README.md
# gomonchickey - go语言实现的工具库
### go version: 1.13

### example code

```go
package main

import (
    "fmt"
    "time"

    "github.com/zengzhiying/gomonchickey"
)

type Coordinate monchickey.Coordinate

func PolygonContain(pointSet []Coordinate, p Coordinate) (int, error) {
    newPointSet := make([]monchickey.Coordinate, len(pointSet))
    for i, c := range pointSet {
        newPointSet[i] = monchickey.Coordinate(c)
    }
    return monchickey.PolygonContain(newPointSet, monchickey.Coordinate(p))
}

func main() {
    nowTimestamp := time.Now().Unix()
    nowTimeStr := monchickey.TimestampToString(nowTimestamp, "2006-01-02 15:04:05")
    fmt.Println(nowTimeStr)

    numSeq := []uint8{72, 101, 108, 108, 111, 32, 109, 111, 110, 99, 104, 105, 99, 107, 101, 121, 33}
    raw := monchickey.Uint8ToBytes(numSeq)
    fmt.Println(string(raw))
    fmt.Println(monchickey.Base64Encode(raw))

    geoHash, err := monchickey.GeohashEncode(113.56291, 36.9271, 12)
    if err != nil {
        fmt.Println(err)
    } else {
        fmt.Println("(113.56291, 36.9271) geohash Encode:", geoHash)
    }
    longitude, latitude, err := monchickey.GeohashDecode(geoHash)
    if err == nil {
        fmt.Println(geoHash, "Decode:(", longitude, latitude, ")")
    }

    pointSet := []Coordinate{
        Coordinate{1, 1},
        Coordinate{1, 4},
        Coordinate{4, 4},
        Coordinate{4, 1},
    }

    fmt.Println("Polygon: (1,1)-(1,4)-(4,4)-(4,1): ")
    v, _ := PolygonContain(pointSet, Coordinate{1, 1})
    fmt.Println("  (1, 1) in", v)  // 边上
    v, _ = PolygonContain(pointSet, Coordinate{2, 2})
    fmt.Println("  (2, 2) in", v)  // 内部
    v, _ = PolygonContain(pointSet, Coordinate{5, 1})
    fmt.Println("  (5, 1) in", v)  // 外部
}


```
> output:  
2020-02-23 15:05:55  
Hello monchickey!  
SGVsbG8gbW9uY2hpY2tleSE=  
(113.56291, 36.9271) geohash Encode: ww8f04mgzw75  
ww8f04mgzw75 Decode:( 113.56290997937322 36.92709996365011 )  
Polygon: (1,1)-(1,4)-(4,4)-(4,1):  
  (1, 1) in 0  
  (2, 2) in 1  
  (5, 1) in -1  


HISTORY.md
# Change HISTORY

### v1.3.0 20201020
> 
 - 扩充float32/64和二进制互转的函数

### v1.2.0 20200401
> 
 - 数字pack/unpack操作函数命名标准化修改
 - 增加单个数字uint64/uint32/uint16的LittleEndian和BigEndian的一系列pack/unpack操作函数
 - 增加区间交集计算和绝对值计算的函数

### v1.1.0 20200223
> 
 - 添加多边形和点的关系计算工具类和测试用例


file_util_test.go
package monchickey_test

import (
    "fmt"
    "testing"

    "time"

    "github.com/zengzhiying/gomonchickey"
)

// go test -v file_util_test.go

func TestPathStatus(t *testing.T) {
    fmt.Println(monchickey.PathIsExist("/usr/aaa"), monchickey.PathIsExist("/usr/bin"))
    fmt.Println(monchickey.PathIsFile("/etc/hosts"), monchickey.PathIsFile("/usr/local/"))
    fmt.Println(monchickey.PathIsDir("/usr/local"), monchickey.PathIsDir("/usr/bin/python"))
}

func TestGobSerialize(t *testing.T) {
    var a = 3
    fmt.Println(monchickey.GobSerialize("/root/a.gob", a))
    a += 2
    fmt.Println(monchickey.GobSerialize("/root/a.gob", a))

    type Aaa struct {
        A int
        B string
    }
    as := Aaa{A: 3, B:"hello"}
    fmt.Println(monchickey.GobSerialize("/root/aaa.gob", as))

    type Bbb struct {
        A time.Time
        B time.Duration
    }
    bb := Bbb{A: time.Now(), B: 10*time.Second}
    fmt.Println(monchickey.GobSerialize("/root/bbb.gob", bb))

    var b int
    err := monchickey.GobDeserialize("/root/a.gob", &b)
    if err != nil {
        fmt.Println(err)
    } else {
        fmt.Println(b)
    }
    aaa := Aaa{}
    err = monchickey.GobDeserialize("/root/aaa.gob", &aaa)
    if err != nil {
        fmt.Println(err)
    } else {
        fmt.Println(aaa)
    }

    bbb := Bbb{}
    err = monchickey.GobDeserialize("/root/bbb.gob", &bbb)
    if err != nil {
        fmt.Println(err)
    } else {
        fmt.Println(bbb)
    }
}


hash_util_test.go
package monchickey_test

import (
    "fmt"
    "testing"

    "github.com/zengzhiying/gomonchickey"
)

// go test -v hash_util_test.go

func TestHashUtil(t *testing.T) {
    message := "This is message string."
    hash64Sum := monchickey.XXHashSum64([]byte(message))
    md5Bytes := monchickey.MD5Digest([]byte(message))
    md5Hex := monchickey.MD5HexDigest([]byte(message))
    fmt.Printf("%d, %q, %s\n", hash64Sum, md5Bytes, md5Hex)
    fmt.Println(fmt.Sprintf("%x", md5Bytes) == md5Hex)
}


config_util_test.go
package monchickey_test

import (
    "fmt"
    "testing"

    "github.com/zengzhiying/gomonchickey"
)

// go test -v config_util_test.go

type Config struct {
    A struct {
        T1 int
        T2 string
    }
}

// go test -v config_util_test.go -test.run TestGetYamlConfig
func TestGetYamlConfig(t *testing.T) {
    config := Config{}
    err := monchickey.GetYamlConfig("resources/config.yaml", &config)
    if err != nil {
        fmt.Println(err)
    } else {
        fmt.Println(config.A.T1, config.A.T2)
        fmt.Println(config.A.T1 == 3, config.A.T2 == "test")
    }
}


file_util.go
package monchickey

import (
    "os"
    "encoding/gob"
)

// 判断文件或目录是不是存在
func PathIsExist(path string) bool {
    _, err := os.Stat(path)
    return err == nil || os.IsExist(err)
}

// 判断path是不是文件
func PathIsFile(path string) bool {
    fi, err := os.Stat(path)
    if err != nil {
        return false
    }
    return !fi.IsDir()
}

// 判断path是不是目录
func PathIsDir(path string) bool {
    fi, err := os.Stat(path)
    if err != nil {
        return false
    }
    return fi.IsDir()
}

// 获得文件的修改时间, 返回字符串示例: 2019-01-01 11:11:00
// 当文件不存在或出错时返回空字符串
func GetFileModifyTime(filename string) string {
    fileInfo, err := os.Stat(filename)
    if err == nil {
        modTime := fileInfo.ModTime()
        return modTime.Format("2006-01-02 15:04:05")
    }
    return ""
}

// 将指定内容以gob方式序列化到文件
// 特殊数据类型需要提前注册
// filePath: 文件绝对路径
func GobSerialize(filePath string, content interface{}) error {
    fp, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
    if err != nil {
        return err
    }
    defer fp.Close()
    encoder := gob.NewEncoder(fp)
    err = encoder.Encode(content)
    if err != nil {
        return err
    }
    return nil
}

// 将gob内容反序列化为数据
// filePath: 文件绝对路径
// content: 接收返回内容的指针
func GobDeserialize(filePath string, content interface{}) error {
    fp, err := os.Open(filePath)
    if err != nil {
        return err
    }
    defer fp.Close()
    decoder := gob.NewDecoder(fp)
    err = decoder.Decode(content)
    if err != nil {
        return err
    }
    return nil
}


data_process.go
package monchickey

import (
    "time"
    "math"
    "errors"
    "encoding/base64"
    "encoding/hex"
    "encoding/binary"
)

// 数据结构常用操作的封装

// 判断整数切片中是否包含元素
func IntArrayContain(arr []int, v int) bool {
    for _, x := range arr {
        if x == v {
            return true
        }
    }
    return false
}

// unix时间戳转时间字符串
// timestamp: 时间戳, 单位: s
// timeFormat: 格式化方式, 例如: "2006-01-02 15:04:05"
func TimestampToString(timestamp int64, timeFormat string) string {
    unixTime := time.Unix(timestamp, 0)
    return unixTime.Format(timeFormat)
}

// 字符串格式的时间转时间戳, 
// 注意: 按照本地时区进行转换
// 返回: 转换成功error为nil, 返回时间戳, 单位为s; 转换失败error不为nil
func StringToTimestamp(timeAsStr, timeFormat string) (int64, error) {
    // UTC
    // ts, err := time.Parse(timeFormat, timeAsStr)
    loc, err := time.LoadLocation("Local")
    if err != nil {
        return 0, err
    }
    ts, err := time.ParseInLocation(timeFormat, timeAsStr, loc)
    if err != nil {
        return 0, err
    }
    return ts.Unix(), nil
}

// 按照指定的时区转换时间戳, 比如: UTC - UTC时间, Asia/Shanghai - 东八区时间
func TimeZoneStringToTimestamp(timeAsStr, timeFormat, timeZone string) (int64, error) {
    loc, err := time.LoadLocation(timeZone)
    if err != nil {
        return 0, err
    }
    ts, err := time.ParseInLocation(timeFormat, timeAsStr, loc)
    if err != nil {
        return 0, err
    }
    return ts.Unix(), nil
}


// 两个字节转float16的数值表示
// 返回值: float64 对应float16表示的浮点数
func ByteToFloat16(high, low byte) float64 {
    s := 1.0
    if high >> 7 == 1 {
        s = -1.0
    }

    exp := int(high & 0x7c) >> 2
    E := exp - 15

    man := int(high & 0x03) << 8 + int(low)
    M := 1.0
    for i := 0; i < 10; i++ {
        v := man & (0x0200 >> uint(i))
        if v != 0 {
            M += 1 / math.Pow(2, float64(i + 1))
        }
    }

    return s * M * math.Pow(2, float64(E))
}

// base64编码
func Base64Encode(data []byte) string {
    return base64.StdEncoding.EncodeToString(data)
}

// base64解码
func Base64Decode(encodedData string) ([]byte, error) {
    return base64.StdEncoding.DecodeString(encodedData)
}

// 校验int是否在uint8范围内
// 校验通过返回value, 否则返回defaultValue默认值
func Uint8Transform(value, defaultValue int) int {
    if value >= 0 && value <= 255 {
        return value
    }
    return defaultValue
}

// 校验int是否在uint16范围内
// 校验通过返回value, 否则返回defaultValue默认值
func Uint16Transform(value, defaultValue int) int {
    if value >= 0 && value <= 65535 {
        return value
    }
    return defaultValue
}

// 编码二进制为hex格式的字符串
func EncodeToHex(src []byte) string {
    return hex.EncodeToString(src)
}

// 解码hex为二进制数组
func HexDecode(s string) ([]byte, error) {
    return hex.DecodeString(s)
}

// uint64数字小端方式打包字节流
func Uint64PackLittleEndian(num uint64) []byte {
    b := make([]byte, 8)
    b[0] = byte(num)
    b[1] = byte(num >> 8)
    b[2] = byte(num >> 16)
    b[3] = byte(num >> 24)
    b[4] = byte(num >> 32)
    b[5] = byte(num >> 40)
    b[6] = byte(num >> 48)
    b[7] = byte(num >> 56)
    return b
}

// 小端方式解包字节数组为uint64数字
func Uint64UnpackLittleEndian(b []byte) uint64 {
    _ = b[7]
    return uint64(b[0]) | uint64(b[1])<<8 | uint64(b[2])<<16 | uint64(b[3])<<24 | 
        uint64(b[4])<<32 | uint64(b[5])<<40 | uint64(b[6])<<48 | uint64(b[7])<<56
}

// uint64数字大端方式打包字节流
func Uint64PackBigEndian(v uint64) []byte {
    b := make([]byte, 8)
    b[0] = byte(v >> 56)
    b[1] = byte(v >> 48)
    b[2] = byte(v >> 40)
    b[3] = byte(v >> 32)
    b[4] = byte(v >> 24)
    b[5] = byte(v >> 16)
    b[6] = byte(v >> 8)
    b[7] = byte(v)
    return b
}

// 从字节数组大端方式解包uint64
func Uint64UnpackBigEndian(b []byte) uint64 {
    _ = b[7]
    return uint64(b[7]) | uint64(b[6])<<8 | uint64(b[5])<<16 | uint64(b[4])<<24 | 
        uint64(b[3])<<32 | uint64(b[2])<<40 | uint64(b[1])<<48 | uint64(b[0])<<56

}

// uint32小端打包
func Uint32PackLittleEndian(v uint32) []byte {
    b := make([]byte, 4)
    b[0] = byte(v)
    b[1] = byte(v >> 8)
    b[2] = byte(v >> 16)
    b[3] = byte(v >> 24)
    return b
}

// uint32小端解包
func Uint32UnpackLittleEndian(b []byte) uint32 {
    _ = b[3]
    return uint32(b[0]) | uint32(b[1])<<8 | uint32(b[2])<<16 | uint32(b[3])<<24
}

// uint32大端打包
func Uint32PackBigEndian(v uint32) []byte {
    b := make([]byte, 4)
    b[0] = byte(v >> 24)
    b[1] = byte(v >> 16)
    b[2] = byte(v >> 8)
    b[3] = byte(v)
    return b
}

// uint32大端解包
func Uint32UnpackBigEndian(b []byte) uint32 {
    _ = b[3]
    return uint32(b[3]) | uint32(b[2])<<8 | uint32(b[1])<<16 | uint32(b[0])<<24
}

// uint16小端打包
func Uint16PackLittleEndian(v uint16) []byte {
    b := make([]byte, 2)
    b[0] = byte(v)
    b[1] = byte(v >> 8)
    return b
}

// uint16小端解包
func Uint16UnpackLittleEndian(b []byte) uint16 {
    _ = b[1]
    return uint16(b[0]) | uint16(b[1])<<8
}

// uint16大端打包
func Uint16PackBigEndian(v uint16) []byte {
    b := make([]byte, 2)
    b[0] = byte(v >> 8)
    b[1] = byte(v)
    return b
}

// uint16大端解包
func Uint16UnpackBigEndian(b []byte) uint16 {
    _ = b[1]
    return uint16(b[1]) | uint16(b[0])<<8
}

// 打包uint8序列为二进制序列
func Uint8ToBytes(seq []uint8) []byte {
    dstBytes := make([]byte, len(seq))
    for i, value := range seq {
        dstBytes[i] = value
    }
    return dstBytes
}

// 打包uint16序列为二进制序列[大端]
func Uint16ToBytesBigEndian(seq []uint16) []byte {
    dstBytes := make([]byte, len(seq) * 2)
    for i, value := range seq {
        high := uint8(value >> 8)
        low := uint8(value)
        dstBytes[2 * i] = high
        dstBytes[2 * i + 1] = low
    }
    return dstBytes
}

// 打包uint16序列为二进制序列[小端]
func Uint16ToBytesLittleEndian(seq []uint16) []byte {
    dstBytes := make([]byte, len(seq) * 2)
    for i , value := range seq {
        high := uint8(value >> 8)
        low := uint8(value)
        dstBytes[2 * i] = low
        dstBytes[2 * i + 1] = high
    }
    return dstBytes
}

// 解包二进制数组为uint16切片[大端]
func BytesToUint16BigEndian(bs []byte) ([]uint16, error) {
    if len(bs) % 2 != 0 {
        return nil, errors.New("Byte length must be a multiple of 2!")
    }
    src := make([]uint16, len(bs) / 2)
    for i := 0; i < len(bs); i += 2 {
        src[i / 2] = uint16(bs[i]) << 8 | uint16(bs[i + 1])
    }
    return src, nil
}

// 解包二进制数组为uint16切片[小端]
func BytesToUint16LittleEndian(bs []byte) ([]uint16, error) {
    if len(bs) % 2 != 0 {
        return nil, errors.New("Byte length must be a multiple of 2!")
    }
    src := make([]uint16, len(bs) / 2)
    for i := 0; i < len(bs); i += 2 {
        src[i / 2] = uint16(bs[i + 1]) << 8 | uint16(bs[i])
    }
    return src, nil
}

// float32小端方式打包为字节数组
func Float32PackLittleEndian(f float32) []byte {
    // 转为uint32
    num := math.Float32bits(f)
    value := make([]byte, 4)
    binary.LittleEndian.PutUint32(value, num)
    return value
}

// float32小端方式解包
func Float32UnpackLittleEndian(v []byte) float32 {
    _ = v[3]
    num := binary.LittleEndian.Uint32(v)
    return math.Float32frombits(num)
}

// float64小端方式打包为字节数组
func Float64PackLittleEndian(f float64) []byte {
    // 转为uint64
    num := math.Float64bits(f)
    value := make([]byte, 8)
    binary.LittleEndian.PutUint64(value, num)
    return value
}

// float64小端方式解包
func Float64UnpackLittleEndian(v []byte) float64 {
    _ = v[7]
    num := binary.LittleEndian.Uint64(v)
    return math.Float64frombits(num)
}

// 设置uint32指定位的值
// index是从高位到低位, 范围: 1 - 32
func SetUint32Bit(num *uint32, index, value uint8) (err error) {
    if index > 32 || index == 0 {
        err = errors.New("Index range is 1 to 32.")
        return
    }
    err = nil
    switch value {
    case 0:
        *num &= ^(1 << (32 - index))
    case 1:
        *num |= 1 << (32 - index)
    default:
        err = errors.New("Value can only be 0 or 1.")
    }
    return
}

// 获取uint32指定位的值, index从高位到低位: 1 - 32
func GetUint32Bit(num uint32, index uint8) (value uint8, err error) {
    if index > 32 || index == 0 {
        err = errors.New("Index range is 1 to 32.")
        return
    }
    err = nil
    movBit := 32 - index
    value = uint8((num & (1 << movBit)) >> movBit)
    return
}

// 设置uint64指定位的值
// index是从高位到低位, index依次为: 1 - 64
func SetUint64Bit(num *uint64, index, value uint8) (err error) {
    if index > 64 || index == 0 {
        err = errors.New("Index range is 1 to 64.")
        return
    }
    err = nil
    switch value {
    case 0:
        *num &= ^(1 << (64 - index))
    case 1:
        *num |= 1 << (64 - index)
    default:
        err = errors.New("Value can only be 0 or 1.")
    }
    return
}

// 获取uint64指定位的值
// index从高位到低位依次为: 1 - 64
func GetUint64Bit(num uint64, index uint8) (value uint8, err error) {
    if index > 64 || index == 0 {
        err = errors.New("Index range is 1 to 64.")
        return
    }
    err = nil
    movBit := 64 - index
    value = uint8((num & (1 << movBit)) >> movBit)
    return
}


// 计算数字的绝对值, int64类型
func Int64Abs(v int64) int64 {
    n := v >> 63
    return (v ^ n) - n
}


// 计算区间[x1, y1]和[x2, y2]的交集区间
// 比如: [-1, 6]与[2, 8]返回: [2, 6], error为nil
// 如果重合在1点, 比如[3,5]与[5,7], 则返回: [5,5]
// 如果区间没有交集, 则error不为nil
func IntervalIntersection(x1, y1, x2, y2 int64) (left, right int64, err error) {
    if x2 <= y1 && x1 <= y2 {
        err = nil
        left, right = x1, y1
        if x2 > x1 {
            left = x2
        }
        if y2 < y1 {
            right = y2
        }
        return
    }
    err = errors.New("No intersection of the two groups.")
    return
}


geohash_util.go
package monchickey

import (
    "fmt"
    "errors"
    "strings"
)

// 经纬度范围
const (
    MAX_LON = 180
    MIN_LON = -180
    MAX_LAT = 90
    MIN_LAT = -90
)

var geoBase32Chars = []string{
    "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
    "b", "c", "d", "e", "f", "g", "h", "j", "k", "m",
    "n", "p", "q", "r", "s", "t", "u", "v", "w", "x",
    "y", "z",
}

var geoCodesByChar = map[byte]uint8{
    '0': 0,
    '1': 1,
    '2': 2,
    '3': 3,
    '4': 4,
    '5': 5,
    '6': 6,
    '7': 7,
    '8': 8,
    '9': 9,
    'b': 10,
    'c': 11,
    'd': 12,
    'e': 13,
    'f': 14,
    'g': 15,
    'h': 16,
    'j': 17,
    'k': 18,
    'm': 19,
    'n': 20,
    'p': 21,
    'q': 22,
    'r': 23,
    's': 24,
    't': 25,
    'u': 26,
    'v': 27,
    'w': 28,
    'x': 29,
    'y': 30,
    'z': 31,
}

// 对单个坐标值进行geohash编码, 位数: digits
// 注意是使用低digits位编码
func coordinateHash(coordinate, min, max float64, digits int) uint32 {
    hashValue := uint32(0)
    for i := 0; i < digits; i++ {
        mid := (min + max) / 2
        if(coordinate >= mid) {
            hashValue |= 1 << (digits - i - 1)
            min = mid
        } else {
            max = mid
        }
    }
    return hashValue
}


// 对单个坐标值做反向hash, 位数: digits
// 注意是使用uint32的高digits位
func coordinateReverseHash(hashValue uint32, min, max float64, digits int) float64 {
    mid := (min + max) / 2
    for i := 1; i <= digits; i++ {
        bitValue, _ := GetUint32Bit(hashValue, uint8(i))
        if bitValue == 1 {
            min = mid
        } else {
            max = mid
        }
        mid = (min + max) / 2
    }
    return mid
}


func base32Encode(src []byte) (string, error) {
    dst := make([]string, len(src))
    for i, c := range src {
        if c > 31 {
            err := errors.New("The source Numbers range from 0 to 31.")
            return "", err
        }
        dst[i] = geoBase32Chars[c]
    }
    return strings.Join(dst, ""), nil
}

func base32Decode(encoded string) ([]byte, error) {
    encodedBytes := []byte(encoded)
    decodedBytes := make([]byte,len(encodedBytes))
    for i, enc := range encodedBytes {
        if dec, ok := geoCodesByChar[enc]; ok {
            decodedBytes[i] = dec
        } else {
            err := errors.New(fmt.Sprintf("Not other characters! %c", enc))
            return decodedBytes, err
        }
    }
    return decodedBytes, nil
}

// geohash编码函数
// longitude: 经度
// latitude: 纬度
// precision: 精度, 范围: 1-12
// returns: geohash编码后的字符串, 错误信息
func GeohashEncode(longitude, latitude float64, precision int) (string, error) {
    if precision < 1 || precision > 12 {
        err := errors.New("Precision range from 1 to 12.")
        return "", err
    }

    // 总位数
    totalDigits := precision * 5
    // 单个坐标值hash位数
    lonDigits := totalDigits / 2
    latDigits := lonDigits
    if totalDigits % 2 != 0 {
        lonDigits++
    }

    lonHash := coordinateHash(longitude, MIN_LON, MAX_LON, lonDigits)
    latHash := coordinateHash(latitude, MIN_LAT, MAX_LAT, latDigits)
    lonHash = lonHash << (32 - lonDigits)
    latHash = latHash << (32 - latDigits)

    mergeHash := uint64(0)
    var i int
    for i = 1; i <= lonDigits; i++ {
        lonbit, _ := GetUint32Bit(lonHash, uint8(i))
        latbit, _ := GetUint32Bit(latHash, uint8(i))
        SetUint64Bit(&mergeHash, uint8(i * 2 - 1), lonbit)
        SetUint64Bit(&mergeHash, uint8(i * 2), latbit)
    }

    if totalDigits % 2 != 0 {
        lonbit, _ := GetUint32Bit(lonHash, uint8(i))
        SetUint64Bit(&mergeHash, uint8(i * 2 - 1), lonbit)
    }

    // 分组转换字节
    hashBytes := make([]byte, precision)
    for i = 1; i <= precision; i++ {
        hashBytes[i - 1] = uint8((mergeHash >> (64 - i * 5)) & 31)
    }

    geoDst, err := base32Encode(hashBytes)

    return geoDst, err
}


// 分割hash为两个坐标的子hash
func splitHash(value uint64, digits int) (uint32, uint32) {
    lonHash := uint32(0)
    latHash := uint32(0)

    var i = 1
    for ; i <= digits - 1; i+=2 {
        lonbit, _ := GetUint64Bit(value, uint8(i))
        latbit, _ := GetUint64Bit(value, uint8(i + 1))
        SetUint32Bit(&lonHash, uint8(i / 2 + 1), lonbit)
        SetUint32Bit(&latHash, uint8(i / 2 + 1), latbit)
    }

    if digits & 0x1 == 1 {
        lonbit, _ := GetUint64Bit(value, uint8(i))
        SetUint32Bit(&lonHash, uint8(i / 2 + 1), lonbit)
    }

    return lonHash, latHash
}

// geohash解码函数
// encoded: hash编码后的字符串, 长度: 1-12
// returns: 经度, 纬度, 错误信息
func GeohashDecode(encoded string) (longitude, latitude float64, err error) {
    precision := len(encoded)
    if precision > 12 || precision == 0 {
        err = errors.New("Encoded length can only be 1 to 12.")
        return
    }

    decoded, err := base32Decode(encoded)
    if err != nil {
        return
    }

    decodedHash := uint64(0)
    for i, dec := range decoded {
        movBit := 64 - (i + 1) * 5
        decodedHash |= (uint64(dec) << movBit)
    }

    lonHash, latHash := splitHash(decodedHash, precision * 5)
    lonDigits := precision * 5 / 2
    latDigits := lonDigits
    if precision & 0x1 == 1 {
        lonDigits++
    }

    longitude = coordinateReverseHash(lonHash, MIN_LON, MAX_LON, lonDigits)
    latitude = coordinateReverseHash(latHash, MIN_LAT, MAX_LAT, latDigits)

    return
}



polygons_point_relation.go
package monchickey

import (
	"errors"
)

// 点和多边形的位置关系计算, 使用射线和多边形的交点数判断点在多边形内部或者外部
// 计算都采用直线的一般式方程：Ax + By + C = 0

type Coordinate struct {
	X int
	Y int
}

// 方程一般式结构体
type GeneralParameter struct {
	A int
	B int
	C int
}

// 给定一点斜率为1的点斜式方程
// (3, 6) => x - y + 3 = 0
func pointSlopeForm(point Coordinate) (int, int, int) {
	return 1, -1, point.Y - point.X
}

// 给定两点计算一般式方程
// 返回0, 0, 0表示两点相同构不成方程
// (-2, -2), (-2, -2) => 0, 0, 0
// (-2, 3), (-2, 6) => x = -2 => x + 2 = 0
// (-2, 3), (-1, 3) => y = 3 => y - 3 = 0
// (-2, 3), (2, 6) => -3x + 4y - 18 = 0
func twoPointForm(p1, p2 Coordinate) (int, int, int) {
	if p1.X == p2.X && p1.Y == p2.Y {
		return 0, 0, 0
	}
	// x = a
	if p1.X == p2.X {
		return 1, 0, -p1.X
	}
	// y = a
	if p1.Y == p2.Y {
		return 0, 1, -p1.Y
	}
	// 一般式
	A := p1.Y - p2.Y
	B := p2.X - p1.X
	C := p1.X * p2.Y - p2.X * p1.Y
	return A, B, C
}

// 解二元一次方程组, 一般式, 根据克勒默法则求解
// 返回err表示方程无解或有无数多个解
func systemOfBinaryLinearQquationsSolving(gp1, gp2 GeneralParameter) (x, y float64, err error) {
	D := gp1.A * gp2.B - gp2.A * gp1.B
	Dx := gp1.B * gp2.C - gp1.C * gp2.B
	Dy := gp1.C * gp2.A - gp2.C * gp1.A
	if D != 0 {
		// 方程有唯一解
		x = float64(Dx) / float64(D)
		y = float64(Dy) / float64(D)
		err = nil
		return
	}
	if Dx != 0 || Dy != 0 {
		// 方程组无解, 两直线平行
		err = errors.New("Equations without solution!")
		return
	}
	// Dx == 0 且 Dy == 0, 两直线重合, 方程有无数个解
	err = errors.New("There are countless solutions to the equation!")
	return
}

// 判断点p是否在gp构成的直线上
// 在返回true, 否则返回false
func lineContain(gp GeneralParameter, p Coordinate) bool {
	return gp.A * p.X + gp.B * p.Y + gp.C == 0
}

// 判断点是否在多边形内
// pointSet: 按照切片的顺序依次连成闭合多边形
// 比如N = 4, 则0->1, 1->2, 2->3, 3->0, 依次连成4个边
// p: 要判断多边形是否包含该点
// p在多边形内返回1, p在多边形边上返回0, 在多边形外返回-1; 同时err为nil
// 当传入错误的数据时, 会返回err
func PolygonContain(pointSet []Coordinate, p Coordinate) (int, error) {
	N := len(pointSet)
	// N个直线方程
	equationSet := make([]GeneralParameter, N)
	for i := 0; i < N - 1; i++ {
		A, B, C := twoPointForm(pointSet[i], pointSet[i + 1])
		if A == 0 && B == 0 && C == 0 {
			return 0, errors.New("Coincident point!")
		}
		equationSet[i] = GeneralParameter{A: A, B: B, C: C}
	}
	A, B, C := twoPointForm(pointSet[N - 1], pointSet[0])
	if A == 0 && B == 0 && C == 0 {
		return 0, errors.New("Coincident point!")
	}
	equationSet[N - 1] = GeneralParameter{A: A, B: B, C: C}

	// 点斜式方程
	pA, pB, pC := pointSlopeForm(p)
	pointSlopeParam := GeneralParameter{A: pA, B: pB, C: pC}
	// 计算点和N个边的交点个数
	numberIntersections := 0
	for i := 0; i < N; i++ {
		// 线段
		lineSeg1 := pointSet[i]
		var lineSeg2 Coordinate
		if i < N - 1 {
			lineSeg2 = pointSet[i + 1]
		} else {
			lineSeg2 = pointSet[0]
		}

		// 边端点的坐标范围
		minX, minY := lineSeg1.X, lineSeg1.Y
		maxX, maxY := lineSeg2.X, lineSeg2.Y
		if minX > maxX {
			minX, maxX = maxX, minX
		}
		if minY > maxY {
			minY, maxY = maxY, minY
		}

		// 判断点是否在多边形边上
		if lineContain(equationSet[i], p) {
			if p.X >= minX && p.X <= maxX && p.Y >= minY && p.Y <= maxY {
				return 0, nil
			}
		}

		// 判断交点是否在射线上
		// 这样的交点算作1个
		if lineContain(pointSlopeParam, pointSet[i]) && pointSet[i].X >= p.X {
			numberIntersections++
			continue
		}

		ix, iy, err := systemOfBinaryLinearQquationsSolving(pointSlopeParam, equationSet[i])
		if err != nil {
			continue
		}
		if ix < float64(p.X) || iy < float64(p.Y) {
			continue
		}

		// 竖直边
		if minX == maxX {
			if iy > float64(minY) && iy < float64(maxY) {
				numberIntersections++
			}
			continue
		}
		// 水平边
		if minY == maxY {
			if ix > float64(minX) && ix < float64(maxX) {
				numberIntersections++
			}
			continue
		}
		// 普通边
		if ix > float64(minX) && ix < float64(maxX) && iy > float64(minY) && iy < float64(maxY) {
			numberIntersections++
		}
	}

	// 内部
	if numberIntersections % 2 != 0 {
		return 1, nil
	}
	// 外部
	return -1, nil
}



examples/monchickey_example.go
package main

import (
    "fmt"
    "time"

    "github.com/zengzhiying/gomonchickey"
)

type Coordinate monchickey.Coordinate

func PolygonContain(pointSet []Coordinate, p Coordinate) (int, error) {
    newPointSet := make([]monchickey.Coordinate, len(pointSet))
    for i, c := range pointSet {
        newPointSet[i] = monchickey.Coordinate(c)
    }
    return monchickey.PolygonContain(newPointSet, monchickey.Coordinate(p))
}

func main() {
    nowTimestamp := time.Now().Unix()
    nowTimeStr := monchickey.TimestampToString(nowTimestamp, "2006-01-02 15:04:05")
    fmt.Println(nowTimeStr)

    numSeq := []uint8{72, 101, 108, 108, 111, 32, 109, 111, 110, 99, 104, 105, 99, 107, 101, 121, 33}
    raw := monchickey.Uint8ToBytes(numSeq)
    fmt.Println(string(raw))
    fmt.Println(monchickey.Base64Encode(raw))

    geoHash, err := monchickey.GeohashEncode(113.56291, 36.9271, 12)
    if err != nil {
        fmt.Println(err)
    } else {
        fmt.Println("(113.56291, 36.9271) geohash Encode:", geoHash)
    }
    longitude, latitude, err := monchickey.GeohashDecode(geoHash)
    if err == nil {
        fmt.Println(geoHash, "Decode:(", longitude, latitude, ")")
    }

    pointSet := []Coordinate{
        Coordinate{1, 1},
        Coordinate{1, 4},
        Coordinate{4, 4},
        Coordinate{4, 1},
    }

    fmt.Println("Polygon: (1,1)-(1,4)-(4,4)-(4,1): ")
    v, _ := PolygonContain(pointSet, Coordinate{1, 1})
    fmt.Println("  (1, 1) in", v)  // 边上
    v, _ = PolygonContain(pointSet, Coordinate{2, 2})
    fmt.Println("  (2, 2) in", v)  // 内部
    v, _ = PolygonContain(pointSet, Coordinate{5, 1})
    fmt.Println("  (5, 1) in", v)  // 外部
}
